[UI-side compositing] Plumb mouseMovedInContentArea event to UI process

nmoucht opened the pull request:
#### 90a47db0ffa461850eaba59ae2b2f2dae80eee2a
<pre>
[UI-side compositing] Plumb mouseMovedInContentArea event to UI process
<a href="https://bugs.webkit.org/show_bug.cgi?id=255126">https://bugs.webkit.org/show_bug.cgi?id=255126</a>
rdar://107732708

Reviewed by Simon Fraser.

This patch plumbs mouseMovedInContentArea events to the UI process.
This event is generated each time the mouse is moved within the
particular content area, and eventually results into a
mouseLocationInScrollerForScrollerImp call into our delegate class,
which allows AppKit to know where the mouse is within the scroller
(this function still needs to be fixed for overflow scroll nodes).
Since there is no state related to this we simply use the property
change on the ScrollingStateNode to tell the UI process to send this
event to the ScrollerPairMac. This patch also ensures that the
NSScrollerImpPairs are accessed on the main thread for the various
mouse events.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::notifyScrollableAreasOfMouseEvents):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setMouseIsActiveInContentArea):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::setMouseIsActiveInContentArea):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::setMouseIsActiveInContentArea):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(WebCore::ScrollerPairMac::mouseMovedInContentArea):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/262699@main">https://commits.webkit.org/262699@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db8349987a1f8a092d228671c7506e91e09e60fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2303 "1 style error") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/2317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2416 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3245 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2330 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2275 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2389 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2072 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2324 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/2029 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2073 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3110 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/18 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2059 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1935 "19 flakes 153 failures") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2133 "Built successfully") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/2059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3247 "run-api-tests-without-change (failure)") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2107 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1899 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2110 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2065 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2041 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/260 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2233 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollingStateNode.h:
```diff
@@ -226,37 +226,38 @@ enum class ScrollingStateNodeProperty : uint64_t {
     VerticalScrollbarLayer                      = 1LLU << 17,
     PainterForScrollbar                         = 1LLU << 18,
     ContentAreaHoverState                       = 1LLU << 19,
+    MouseActivtyState                           = 1LLU << 20,
```
Spelling

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -285,6 +285,11 @@ void ScrollingStateScrollingNode::setMouseIsOverContentArea(bool flag)
 
+void ScrollingStateScrollingNode::setMouseIsActiveInContentArea()
+{
+    setPropertyChanged(Property::MouseActivtyState);
```
It’s weird to have a state bit but with no associated state. Previously we’ve had bugs when the scrolling state tree has changes that aren’t stateful (older versions of setRequestedScrollData were like this). But this seems OK for now.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -351,21 +351,36 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
+    });
 }
 
+void ScrollerPairMac::mouseMovedInContentArea()
```
Why not use the “mouse move” terminology in the state tree?

nmoucht replied:
Fixed.

nmoucht committed:
[UI-side compositing] Plumb mouseMovedInContentArea event to UI process
https://bugs.webkit.org/show_bug.cgi?id=255126
rdar://107732708

Reviewed by Simon Fraser.

This patch plumbs mouseMovedInContentArea events to the UI process.
This event is generated each time the mouse is moved within the
particular content area, and eventually results into a
mouseLocationInScrollerForScrollerImp call into our delegate class,
which allows AppKit to know where the mouse is within the scroller
(this function still needs to be fixed for overflow scroll nodes).
Since there is no state related to this we simply use the property
change on the ScrollingStateNode to tell the UI process to send this
event to the ScrollerPairMac. This patch also ensures that the
NSScrollerImpPairs are accessed on the main thread for the various
mouse events.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::notifyScrollableAreasOfMouseEvents):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setMouseIsActiveInContentArea):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::setMouseIsActiveInContentArea):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::setMouseIsActiveInContentArea):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(WebCore::ScrollerPairMac::mouseMovedInContentArea):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/262699@main
